File: src/day01.zig
const std = @import("std");
pub const std_options: std.Options = .{ .log_level = .info };

pub fn main() !void {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var args = try std.process.argsWithAllocator(allocator);
    defer args.deinit();

    // skip exectutable name
    _ = args.skip();

    const filename = args.next();
    if (filename == null) {
        std.log.err("Missing filename", .{});
        return;
    }

    const tic = std.time.microTimestamp();
    var part1: u64 = 0;
    var part2: u64 = 0;
    const nrun = 10000;
    for (0..nrun) |_| {
        var file = try std.fs.cwd().openFile(filename.?, .{ .mode = .read_only });
        defer file.close();

        const read_buf = try file.readToEndAlloc(allocator, 1024 * 1024);
        defer allocator.free(read_buf);
        var it = std.mem.splitAny(u8, read_buf, "\n");

        var acc_part1: u64 = 0;
        var acc_part2: u64 = 0;

        while (it.next()) |line| {
            // part1
            {
                var firstDigit: u32 = 0;
                var lastDigit: u32 = 0;
                forward: for (0..line.len) |idx| {
                    const digit = line[idx] - '0';
                    if (digit >= 0 and digit < 10) {
                        firstDigit = digit;
                        break :forward;
                    }
                }
                backward: for (0..line.len) |idx| {
                    const digit = line[line.len - idx - 1] - '0';
                    if (digit >= 0 and digit < 10) {
                        lastDigit = digit;
                        break :backward;
                    }
                }
                acc_part1 += firstDigit * 10 + lastDigit;
            }
            // part2
            {
                const digits_string = [_][]const u8{ "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };
                var firstDigit: u64 = 0;
                var lastDigit: u64 = 0;
                // We use this method because word can overlap like "nineight"
                forward: for (0..line.len) |idx| {
                    const digit = line[idx] - '0';
                    if (digit >= 0 and digit < 10) {
                        firstDigit = digit;
                        break :forward;
                    } else {
                        const slice = line[idx..];
                        for (digits_string, 1..) |str, value| {
                            if (std.mem.startsWith(u8, slice, str)) {
                                firstDigit = value;
                                break :forward;
                            }
                        }
                    }
                }
                backward: for (0..line.len) |idx| {
                    const ridx = line.len - idx - 1;
                    const digit = line[ridx] - '0';
                    if (digit >= 0 and digit < 10) {
                        lastDigit = digit;
                        break :backward;
                    } else {
                        const slice = line[ridx..];
                        for (digits_string, 1..) |str, value| {
                            if (std.mem.startsWith(u8, slice, str)) {
                                lastDigit = value;
                                break :backward;
                            }
                        }
                    }
                }
                acc_part2 += firstDigit * 10 + lastDigit;
            }
        }
        part1 = acc_part1;
        part2 = acc_part2;
    }
    const tac: i64 = std.time.microTimestamp() - tic;
    std.log.info("Zig  day01 in {d:>20.2} us : part1={:<10} part2={:<10}", .{ @as(f32, @floatFromInt(tac)) / @as(f32, nrun), part1, part2 });
}

File: src/day01.rs
use std::time::Instant;
use std::iter::zip;

type Result<T> = ::std::result::Result<T, Box<dyn (::std::error::Error)>>;

fn main() -> Result<()> {
  let args: Vec<String> = std::env::args().collect();
  if args.len() <= 1 {
    Err("Missing filename")?;
  }
  let filename = &args[1];

  let now = Instant::now();

  let mut part1: u64 = 0;
  let mut part2: u64 = 0;
  let nrun = 10000;

  for _ in 0..nrun {
    let mut digits1: Vec<i32> = Vec::new();
    let mut digits2: Vec<u64> = Vec::new();
    for line in std::fs::read_to_string(filename)?.lines() {
      let line_str = line;
      // part1
      let parsed_line: Vec<i32> = line_str
        .as_bytes()
        .iter()
        .map(|c| *c as i32 - '0' as i32)
        .filter(|d| *d >= 0 && *d < 10i32)
        .collect::<Vec<_>>();
      let parsed_len = parsed_line.len();
      if parsed_len > 0 {
        digits1.push(parsed_line[0] * 10 + parsed_line[parsed_len - 1]);
      }
      // part2
      let mut parsed_line: Vec<u64> = vec![];
      let line_len = line_str.len();
      // We use this method because word can overlap like "nineight"
      let digits_string = ["one", "two", "three", "four", "five", "six", "seven", "eight", "nine"];
      'outer: for i in 0..line_len {
        let slice = &line_str[i..];
        if let Some(d) = slice.chars().nth(0).ok_or("Failed to get first digit")?.to_digit(10) {
          parsed_line.push(d as u64);
        } else {
          for (num, str) in zip(1.., digits_string.iter()) {
            if slice.starts_with(str) {
              parsed_line.push(num);
              continue 'outer;
            }
          }
        }
      }
      let parsed_len = parsed_line.len();
      digits2.push(parsed_line[0] * 10 + parsed_line[parsed_len - 1]);
    }
    part1 = digits1.iter().sum::<i32>() as u64;
    part2 = digits2.iter().sum();
  }
  let duration = now.elapsed().as_micros();
  println!(
    "info: Rust day01 in {:>20.2} us : part1={:<10} part2={:<10}",
    duration as f32 / nrun as f32,
    part1,
    part2
  );

  Ok(())
}
